pkg/operator/k8sutil/job.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package k8sutil for Kubernetes helpers.
+package k8sutil
+
+import (
+	"fmt"
+	"time"
+
+	batch "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+func RunReplaceableJob(clientset kubernetes.Interface, job *batch.Job) error {
+	// check if the job was already created and what its status is
+	existingJob, err := clientset.Batch().Jobs(job.Namespace).Get(job.Name, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		logger.Warningf("failed to detect job %s. %+v", job.Name, err)
+	} else if err == nil {
+		// if the job is still running, allow it to continue to completion
+		if existingJob.Status.Active > 0 {
+			logger.Infof("Found previous job %s. Status=%+v", job.Name, existingJob.Status)
+			return nil
+		}
+
+		// delete the job that already exists from a previous run
+		logger.Infof("Removing previous job %s to start a new one", job.Name)
+		err := DeleteBatchJob(clientset, job.Namespace, existingJob.Name, true)
+		if err != nil {
+			logger.Warningf("failed to remove job %s. %+v", job.Name, err)
+		}
+	}
+
+	_, err = clientset.Batch().Jobs(job.Namespace).Create(job)
+	return err
+}
+
+// WaitForJobCompletion waits for a job to reach the completed state
+func WaitForJobCompletion(clientset kubernetes.Interface, job *batch.Job) error {
+	logger.Infof("waiting for job %s to complete...", job.Name)
+	wait.Poll(5*time.Second, 15*time.Minute, func() (bool, error) {
+		job, err := clientset.Batch().Jobs(job.Namespace).Get(job.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("failed to detect job %s. %+v", job.Name, err)
+		}
+
+		// if the job is still running, allow it to continue to completion
+		if job.Status.Active > 0 {
+			logger.Debugf("job is still running. Status=%+v", job.Status)
+			return false, nil
+		}
+		if job.Status.Failed > 0 {
+			return false, fmt.Errorf("job %s failed.", job.Name)
+		}
+		if job.Status.Succeeded > 0 {
```

Comment:
is this assuming that there will only be a single job that is running and as soon as that 1 job finishes we call the whole thing done?

Reply:
This assumption is valid in the case of the version check, but since this is in the `k8sutil` package I need to make this more general.